### DISCUSSION TITLE
How can I use UART port on MKS Tinybee?

### DISCUSSION THREAD

**Discussion by ashokmanker** [grblHAL/core#456] [Q&A]
How can I use UART port on MKS Tinybee. Its UART2 connected on pin16 and pin17 of esp32. In FluidNC i have defined in config file and its working, but how can I do in grblhal?

**Comment by terjeio** [comment]
Use for what purpose? In your own plugin?

**Reply by ashokmanker** [reply]
Thank you so much for such a prompt reply. I want to use it for my offline controller. Its made with esp32 and 3.5inch tft. It communicate with uart needs rx,tx,5v,gnd. In FluidNC I defined UART2 and assigned pins and defined UART channel for this port.

**Reply by terjeio** [reply]
[Uncomment this line](https://github.com/grblHAL/ESP32/blob/fa7b9196ec64ad6bec013a3ff3d5d29f24ba94e4/main/my_machine.h#L67) and change the value to 2.
Note that grblHAL does not allow mingling gocde/commands from several inputs. If you want full control via the UART you'll have to request that by sending the `CMD_MPG_MODE_TOGGLE` character from the offline controller. More [here](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces).

**Reply by ashokmanker** [reply]
Thanks a lot. I will try it. And report it back.

**Reply by ashokmanker** [reply]
I have tried as per your suggestion. Now offline controller sees timybee board connected. Can display machine cordinates. But cant send commands to tinybee cnc controller. I don't know how to send 0x8B from controller. It can send only grbl codes. Tried sending $0x8B from controller and from console. I have to unplug USB from PC while using offline controller.

**Reply by terjeio** [reply]
0x8B is a single character, decimal value 139 - it has no keyboard representation that I know of.
If you send the real-time query character ? like this: `serial_send_char('?')` then `serial_send_char(0x8B)` is how to send it (no quotes).

**Reply by ashokmanker** [reply]
0x8B is "<" char and its alt+139. Tried sending by keyboard but still not able to use it. Dont know any other way. Can i configure grblhal that it uses uart2 only. Which i am connected to offline controller. When ever I need pc connectivity i will unplug controller and use usb or via Wi-Fi

**Reply by terjeio** [reply]
>  Can i configure grblhal that it uses uart2 only. 

Try adding `#define SERIAL_STREAM 1` to _my_machine.h_.

**Reply by ashokmanker** [reply]
Serial_stream 1 causes cpu fault

**Reply by terjeio** [reply]
Attempted fix has now been committed. If this does not help more details are required.

**Reply by ashokmanker** [reply]
Same problem

**Reply by terjeio** [reply]
Try replacing [this code](https://github.com/grblHAL/ESP32/blob/92977c3cb07f4d604597d014228f4aaf57b5f5f6/main/driver.h#L248-L262) with:

```
#ifdef DEBUGOUT
#define DEBUG_TEST 1
#else
#define DEBUG_TEST 0
#endif

#if (MODBUS_TEST + KEYPAD_TEST + MPG_TEST + TRINAMIC_TEST + DEBUG_TEST) > (SP0 + SP1 + SP2)
#error "Too many options that requires a serial port are enabled!"
#elif (MODBUS_TEST + KEYPAD_TEST + MPG_TEST + TRINAMIC_TEST + DEBUG_TEST) || SERIAL_STREAM == 1
#define SERIAL2_ENABLE 1
#else
#define SERIAL2_ENABLE 0
#endif

#undef SP0
#undef SP1
#undef SP2
#undef MODBUS_TEST
#undef KEYPAD_TEST
#undef MPG_TEST
#undef TRINAMIC_TEST
#undef DEBUG_TEST
```
